### PR TITLE
[expo-media-library][ios] Add `presentPermissionsPickerAsync`

### PR DESCRIPTION
**Pull request by lukmccall**
# Why

Add function (see https://github.com/expo/expo/pull/9423#issuecomment-707237822) which will display prompt with permitted assets. 
image

# How

- added `presentLimitedLibraryPickerAsync` function 
- added example to NCL 
- added a way to check if the user changed the permitted assets 

# ToDo: 

- [x] run pod install 
  - [x] client
  - [x] bare-expo

# Test Plan

- NCL:
  - bare ✅
  - client ✅

# Changelog

[feat] Added `MediaLibrary.presentLimitedLibraryPickerAsync` function, which will display the system prompt allowing the users to change the selected permitted assets` on iOS.

### CODE REVIEW

File: packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m [hunk 1]
```diff
@@ -532,12 +547,13 @@ - (void)photoLibraryDidChange:(PHChange *)changeInstance
       _allAssetsFetchResult = changeDetails.fetchResultAfterChanges;
       
       // PHPhotoLibraryChangeObserver is calling this method too often, so we need to filter out some calls before they are sent to JS.
-      // Ultimately, we emit an event only when something has been inserted or removed from the library.
+      // Ultimately, we emit an event when something has been inserted, removed from the library or changed the permissions.
```
Comment (bbarthec) [inline packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m:550]:
```suggestion
      // Ultimately, we emit an event when something has been inserted or removed from the library, or the user changed the permissions.
```

File: packages/expo-media-library/src/MediaLibrary.ts [hunk 2]
```diff
@@ -192,6 +193,13 @@ export async function getPermissionsAsync(): Promise<PermissionResponse> {
   return await MediaLibrary.getPermissionsAsync();
 }
 
+export async function presentLimitedLibraryPickerAsync(): Promise<void> {
+  if (MediaLibrary.presentLimitedLibraryPickerAsync) {
+    return await MediaLibrary.presentLimitedLibraryPickerAsync();
+  }
+  return Promise.resolve();
+}
```
Comment (bbarthec) [inline packages/expo-media-library/src/MediaLibrary.ts:201]:
```suggestion
/**
 * @iOS-only
 * @throws Will throw an error if called on platform that doesn't support this functionality (eg. iOS < 14, Android, etc.).
 // or
 * @returns `true` when the picker is successfully presented and `false` otherwise (eg. iOS < 14, Android, etc.).
 **/
export async function presentLimitedLibraryPickerAsync(): Promise<void> {
  if (MediaLibrary.presentLimitedLibraryPickerAsync) {
    return await MediaLibrary.presentLimitedLibraryPickerAsync();
  }
  return false;
}
```
Isn't using `async` keyword in function signature auto-magically wrap the return value in a `Promise`? 🤔 
I'm not sure if resolving with the same value `void` when nothing happens (on platforms that not support this function) is a good idea 🤔 
What about resolving to `true` if the picker is presented (on iOS 14+) and `false` otherwise (iOS < 14, Android, etc.)?
With the way it's now we have two failing scenario, but one is not reported at all:
- `a)` iOS < 14 - it will throw from the native code
- `b)` no native method (eg. Android) - results as if the modal was presented 🤔 

Other idea it to throw on every platform that is not supporting this functionality 🤔 I'm not sure which approach will be better, but we tend to throw if sth is not supported in other places, so maybe that's the way 🤔 🤷 

cc @cruzach @esamelson

File: docs/pages/versions/unversioned/sdk/media-library.md [hunk 3]
```diff
@@ -230,6 +240,7 @@ Subscribes for updates in user's media library.
 #### Arguments
 
 - **listener (_function_)** -- A callback that is called when any assets have been inserted or deleted from the library. **On Android** it's invoked with an empty object. **On iOS** it's invoked with an object that contains following keys:
+  - **shouldFullyReload (_boolean_)** -- Whether to all assets reloads is needed. When the user changes the permissions, this value will be set to `true` and you should reload all used assets.
```
Comment (bbarthec) [inline docs/pages/versions/unversioned/sdk/media-library.md:243]: This name is very confusing without reading the whole docs and it suggest specific scenario that should be performed 😅 I'm not a fan of this 😉 

Let's see what is actually happening. The idea is to inform that the selection has changed yes? So what about naming it that way?
Names coming to my mind are: `assetsPermissionsChanged`, `permittedAssetsSelectionChanged` 🤔 

```suggestion
  - **assetsPermissionsChanged (_boolean_)** -- Flag indicating the permissions change in the `user's limited library selection`. If it's `true` this means that the user changed the access to the individual assets in the library. If this happens you should reload all used assets to resemble the actual access permissions.
```

cc @cruzach

File: packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m [hunk 4]
```diff
@@ -554,6 +570,19 @@ - (void)photoLibraryDidChange:(PHChange *)changeInstance
...
+      }
+      
+      if (@available(ios 14, *)) {
+        // Emit event when the user changed the limited permissions.
+        if (!changeDetails.hasIncrementalChanges) {
+          [_eventEmitter sendEventWithName:EXMediaLibraryDidChangeEvent body:@{
+            @"shouldFullyReload": @(true),
+            @"insertedAssets": @[],
+            @"deletedAssets": @[],
+            @"updatedAssets": @[]
```
Comment (bbarthec) [inline packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m:583]: I know this might be confusing, but what about splitting this into the union type:
```
| {
    insertedAssets: Assets[],
    deletedAssets: Assets[],
    updatedAssets: Assets[]
  }
| {
    shouldFullyReload: boolean // this name should change
  }
 
```

**Review by bbarthec — CHANGES_REQUESTED** [review]
I've got a couple of thoughts on this API 😉

File: docs/pages/versions/unversioned/sdk/media-library.md (hunk 3 above)
Comment (cruzach) [reply docs/pages/versions/unversioned/sdk/media-library.md:243]: I agree, `assetsPermissionsChanged` makes it clear what's happening behind the scenes, rather than just us telling the user what to do (fully reload)
```suggestion
  - ** assetPermissionsChanged (_boolean_)** -- Flag indicating a permissions change occurred in the user's limited library selection. If `true`, the user has changed access to individual assets in their library, and you should reload all used assets to comply with the new permissions.
```

File: docs/pages/versions/unversioned/sdk/media-library.md [hunk 5]
```diff
@@ -230,6 +240,7 @@ Subscribes for updates in user's media library.
 #### Arguments
 
 - **listener (_function_)** -- A callback that is called when any assets have been inserted or deleted from the library. **On Android** it's invoked with an empty object. **On iOS** it's invoked with an object that contains following keys:
```
Comment (cruzach) [inline docs/pages/versions/unversioned/sdk/media-library.md:242]: We should updated this description now that the listener is also called when the permissions change. Maybe something like
```suggestion
- **listener (_function_)** -- A callback that is fired when any assets have been inserted or deleted from the library, or when the user changes which assets they're allowing access to. **On Android** it's invoked with an empty object. **On iOS** it's invoked with an object containing following keys:
```

File: packages/expo-media-library/src/MediaLibrary.ts (hunk 2 above)
Comment (lukmccall) [reply packages/expo-media-library/src/MediaLibrary.ts]: I will stick with rejecting the promise if the method is not available.

File: packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m (hunk 4 above)
Comment (lukmccall) [reply packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m]: I'm not a big fan of this. However, I split it to:
```
  /**
   * iOS only
   */
  | {
      assetPermissionsChanged: boolean;
      insertedAssets: Asset[];
      deletedAssets: Asset[];
      updatedAssets: Asset[];
    }
  | object;
```

File: docs/pages/versions/unversioned/sdk/media-library.md [hunk 6]
```diff
@@ -229,7 +239,8 @@ Subscribes for updates in user's media library.
 
 #### Arguments
 
-- **listener (_function_)** -- A callback that is called when any assets have been inserted or deleted from the library. **On Android** it's invoked with an empty object. **On iOS** it's invoked with an object that contains following keys:
+- **listener (_function_)** -- A callback that is fired when any assets have been inserted or deleted from the library, or when the user changes which assets they're allowing access to. **On Android** it's invoked with an empty object. **On iOS** it's invoked with an object containing following keys:
+  - **assetPermissionsChanged (_boolean_)** -- Flag indicating a permissions change occurred in the user's limited library selection. If `true`, the user has changed access to individual assets in their library, and you should reload all used assets to comply with the new permissions.
```
Comment (bbarthec) [inline docs/pages/versions/unversioned/sdk/media-library.md:243]:
```suggestion
  - **assetsPermissionsChanged (_boolean_)** -- Flag indicating a permissions change occurred in the user's limited library selection. If `true`, the user has changed access to individual assets in their library, and you should reload all used assets to comply with the new permissions.
```
Yeah, I like this name more then the previous one 😉 
I suggest going with the plural form as there's number of assets rather than single one 🤔

File: packages/expo-media-library/src/MediaLibrary.ts [hunk 7]
```diff
@@ -69,11 +69,17 @@ export type MediaLibraryAssetInfoQueryOptions = {
...
+  /**
+   * iOS only
+   */
+  | {
+      assetPermissionsChanged: boolean;
+      insertedAssets: Asset[];
+      deletedAssets: Asset[];
+      updatedAssets: Asset[];
+    }
+  | object;
```
Comment (bbarthec) [inline packages/expo-media-library/src/MediaLibrary.ts:82]: I don't really get the idea of adding `| object` to this EventType? 🤔
I suggest going with the union type:
```
export type MediaLibraryAssetChangeEvent =
  | {
      assetPermissionsChanged: boolean;
    }
  | {
      insertedAssets: Asset[];
      deletedAssets: Asset[];
      updatedAssets: Asset[];
    }
```
And a side note: it should be `MediaLibraryAssetsChangeEvent` and not `MediaLibrary Asset ChangeEvent` as this event reports the change in multiple assets 🤔 
Do you mind creating a separate PR fixing this name? 😅

File: packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m [hunk 8]
```diff
@@ -566,6 +586,19 @@ - (void)photoLibraryDidChange:(PHChange *)changeInstance
...
+      }
+      
+      if (@available(ios 14, *)) {
+        // Emit event when the user changed the limited permissions.
+        if (!changeDetails.hasIncrementalChanges) {
+          [_eventEmitter sendEventWithName:EXMediaLibraryDidChangeEvent body:@{
+            @"assetPermissionsChanged": @(true),
+            @"insertedAssets": @[],
+            @"deletedAssets": @[],
+            @"updatedAssets": @[]
```
Comment (bbarthec) [inline packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m:599]:
```suggestion
```

File: packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m [hunk 9]
```diff
@@ -544,12 +563,13 @@ - (void)photoLibraryDidChange:(PHChange *)changeInstance
...
       
       // PHPhotoLibraryChangeObserver is calling this method too often, so we need to filter out some calls before they are sent to JS.
-      // Ultimately, we emit an event only when something has been inserted or removed from the library.
+      // Ultimately, we emit an event when something has been inserted or removed from the library, or the user changed the permissions.
       if (changeDetails.hasIncrementalChanges && (changeDetails.insertedObjects.count > 0 || changeDetails.removedObjects.count > 0)) {
         NSMutableArray *insertedAssets = [NSMutableArray new];
         NSMutableArray *deletedAssets = [NSMutableArray new];
         NSMutableArray *updatedAssets = [NSMutableArray new];
         NSDictionary *body = @{
+                               @"assetPermissionsChanged": @(false),
```
Comment (bbarthec) [inline packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m:572]:
```suggestion
```

**Review by bbarthec — CHANGES_REQUESTED** [review]
Ok, I think we failed to understand each other in regards to https://github.com/expo/expo/pull/10662#discussion_r504538406 😅 

I read in the code there's either an `AssetChangeEvent` (it should be `AssetsChangeEvent` anyway) reporting some `Assets` are `added/deleted/updated` or this other event that only reports that the permissions are updated. Why not to diverge these two events in the TS type system as well? 🤔

File: packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m [hunk 10]
```diff
@@ -128,6 +129,24 @@ - (id)requesterClass:(BOOL)writeOnly
                                                                 reject:reject];
 }
 
+UM_EXPORT_METHOD_AS(presentLimitedLibraryPickerAsync,
```
Comment (tsapeta) [inline packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m:132]: I know this name is taken from `PHPhotoLibrary`'s method that you call, but it really sucks — it's kinda confusing in what it does. What about `presentPermissionsPickerAsync`? Or even making it a part of `requestPermissionsAsync` if the current permission is set to limited?

**Review by tsapeta — COMMENTED** [review]
Left a comment below. Also, have you checked what happens when this native method is called and the permission is already given for all assets? I'm just curious but it might be worth adding to the docs too 😄

File: docs/pages/versions/unversioned/sdk/media-library.md [hunk 11]
```diff
@@ -49,6 +49,16 @@ Checks user's permissions for accessing media library. Alias for `Permissions.ge
 
 A promise that resolves to an object of type [CameraRollPermissionResponse](#medialibrarycamerarollpermissionresponse).
 
+### `MediaLibrary.presentLimitedLibraryPickerAsync()`
+
+**Available only on iOS >= 14.** Allows the user to update the assets which your app will have access to. This method will display a system prompt only if permissions were `granted` and the `accessPrivileges` is equal to `limited`.
```
Comment (tsapeta) [inline docs/pages/versions/unversioned/sdk/media-library.md:54]:
```suggestion
**Available only on iOS >= 14.** Allows the user to update the assets which your app has access to. It displays a system modal only if permissions were `granted` and the `accessPrivileges` is equal to `limited`.
```

File: packages/expo-media-library/CHANGELOG.md [hunk 12]
```diff
@@ -4,8 +4,12 @@
 
 ### 🛠 Breaking changes
 
+- Renamed `MediaLibrary.MediaLibraryAssetChangeEvent` type to `MediaLibrary.MediaLibraryAssetChangeEvent`.
```
Comment (bbarthec) [inline packages/expo-media-library/CHANGELOG.md:7]:
```suggestion
- Renamed `MediaLibrary.MediaLibraryAssetChangeEvent` type to `MediaLibrary.MediaLibraryAssetsChangeEvent`.
```

File: packages/expo-media-library/CHANGELOG.md [hunk 13]
```diff
@@ -4,8 +4,12 @@
...
+
 ### 🎉 New features
 
+- Added the `MediaLibrary.presentLimitedLibraryPickerAsync` function, which will display the system prompt allowing the users to change the selected permitted assets` on iOS.
```
Comment (bbarthec) [inline packages/expo-media-library/CHANGELOG.md:11]:
```suggestion
- Added the `MediaLibrary.presentPermissionsPickerAsync` method that displays the system prompt allowing the user to change the selected permitted assets` on iOS.
```

File: docs/pages/versions/unversioned/sdk/media-library.md [hunk 14]
```diff
@@ -42,9 +42,9 @@ Checks user's permissions for accessing media library. Alias for `Permissions.ge
 
 A promise that resolves to an object of type [CameraRollPermissionResponse](#medialibrarycamerarollpermissionresponse).
 
-### `MediaLibrary.presentLimitedLibraryPickerAsync()`
+### `MediaLibrary.presentPermissionsPickerAsync()`
 
-**Available only on iOS >= 14.** Allows the user to update the assets which your app will have access to. This method will display a system prompt only if permissions were `granted` and the `accessPrivileges` is equal to `limited`.
+**Available only on iOS >= 14.** Allows the user to update the assets which your app has access to. It displays a system modal only if permissions were `granted` and the `accessPrivileges` is equal to `limited`.
```
Comment (bbarthec) [inline docs/pages/versions/unversioned/sdk/media-library.md:47]:
```suggestion
**Available only on iOS >= 14.** Allows the user to update the assets which your app has access to. The system modal is displayed only if a user originally selected the `limited` access to the media library.
```
I would go with sth like this 🤔 
I guess there's still a better and simpler way to describe the actual behaviour than what I'm suggesting 😅

File: docs/pages/versions/unversioned/sdk/media-library.md [hunk 15]
```diff
@@ -233,7 +233,11 @@ Subscribes for updates in user's media library.
 #### Arguments
 
 - **listener (_function_)** -- A callback that is fired when any assets have been inserted or deleted from the library, or when the user changes which assets they're allowing access to. **On Android** it's invoked with an empty object. **On iOS** it's invoked with an object containing following keys:
-  - **assetPermissionsChanged (_boolean_)** -- Flag indicating a permissions change occurred in the user's limited library selection. If `true`, the user has changed access to individual assets in their library, and you should reload all used assets to comply with the new permissions.
+
+  - **hasIncrementalChanges (_boolean_)** -- Whether the changes could be described by the incremental changes. `true` if the changes are described by the removed/inserted/changed details. `false` indicates that the scope of changes were too large and UI clients should do a full reload, incremental changes could not be provided (for example, if the user has changed access to individual assets in their library).
```
Comment (bbarthec) [inline docs/pages/versions/unversioned/sdk/media-library.md:237]:
```suggestion
  - **hasIncrementalChanges (_boolean_)** -- Whether the media library's changes could be described by the `incremental changes`. `true` indicates the changes are described by the `insertedAssets`, `deletedAssets` and `updatedAssets` values. `false` indicates that the scope of changes is too large and a full assets reload should be performed by the application, `incremental changes` could not be provided in this case (eg. a user has changed the access to the individual assets in the media library).
```

File: docs/pages/versions/unversioned/sdk/media-library.md [hunk 16]
```diff
@@ -233,7 +233,11 @@ Subscribes for updates in user's media library.
...
+
+  Presents only if `hasIncrementalChanges` is `true`:
```
Comment (bbarthec) [inline docs/pages/versions/unversioned/sdk/media-library.md:239]:
```suggestion
  Available only if `hasIncrementalChanges` is `true`:
```

**Review by bbarthec — COMMENTED** [review]
A few language suggestions 😉

**Review by bbarthec — APPROVED** [review]
Looks better now 😉 
I approve, but also summon @cruzach for another docs language check 🧙‍♂️

File: docs/pages/versions/unversioned/sdk/media-library.md [hunk 17]
```diff
@@ -222,7 +232,11 @@ Subscribes for updates in user's media library.
 
 #### Arguments
 
-- **listener (_function_)** -- A callback that is called when any assets have been inserted or deleted from the library. **On Android** it's invoked with an empty object. **On iOS** it's invoked with an object that contains following keys:
+- **listener (_function_)** -- A callback that is fired when any assets have been inserted or deleted from the library, or when the user changes which assets they're allowing access to. **On Android** it's invoked with an empty object. **On iOS** it's invoked with an object containing following keys:
+  - **hasIncrementalChanges (_boolean_)** -- Whether the media library's changes could be described by the `incremental changes`. `true` indicates the changes are described by the `insertedAssets`, `deletedAssets` and `updatedAssets` values. `false` indicates that the scope of changes is too large and a full assets reload should be performed by the application, `incremental changes` could not be provided in this case (eg. a user has changed the access to the individual assets in the media library).
```
[the remark this answers is not in the document]
Comment (lukmccall) [reply docs/pages/versions/unversioned/sdk/media-library.md:236]: So when the user change permissions the `hasIncrementalChanges` key will be set to false. This is an example that shows when the user needs to reload all assets.

File: docs/pages/versions/unversioned/sdk/media-library.md [hunk 18]
```diff
@@ -42,6 +42,16 @@ Checks user's permissions for accessing media library. Alias for `Permissions.ge
...
 
+### `MediaLibrary.presentPermissionsPickerAsync()`
+
+**Available only on iOS >= 14.** Allows the user to update the assets which your app has access to. The system modal is displayed only if a user originally selected the `limited` access to the media library.
+
+#### Returns
+
+A promise that resolves to void if prompt will be displayed correctly.
+
+> **Note:** This method doesn't inform you if the user changes permitted assets. To check this you need to subscribe for updates in the user's media library using [MediaLibrary.addListener(listener)](#medialibraryaddlistenerlistener). If `shouldFullyReload` is set to true, the user changed the permissions.
```
[the remark this answers is not in the document]
Comment (lukmccall) [reply docs/pages/versions/unversioned/sdk/media-library.md:53]:
```suggestion
> **Note:** This method doesn't inform you if the user changes which assets your app has access to. For that information, you need to subscribe for updates to the user's media library using [MediaLibrary.addListener(listener)](#medialibraryaddlistenerlistener). If `hasIncrementalChanges` is `false`, the user changed their permissions.
```

Ohh. I think this should look like this :D 

If the user has changed permissions the `hasIncrementalChanges` will be set to `false`.

[the remark this answers is not in the document]
Comment (cruzach) [reply docs/pages/versions/unversioned/sdk/media-library.md]: aha! okay that makes more sense 😄

File: docs/pages/versions/unversioned/sdk/media-library.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (cruzach) [reply docs/pages/versions/unversioned/sdk/media-library.md]: Got it, I understand now. With [this change](https://github.com/expo/expo/pull/10662#discussion_r522863842), that will be clear now in the docs 👍

**Comment by Ashoat** [comment]
Thanks for implementing this functionality!

I'd like to get some advice. I have a media gallery implemented as a custom keyboard, and I am trying to figure out a way to make it work with iOS 14's new "limited library" functionality.

My issue is that I can't get the keyboard to show at the same time as the system's "limited library picker". If the keyboard is already open when the picker gets shown, the keyboard gets dismissed. If the keyboard is shown while the picker is shown, it gets displayed over the picker, which isn't what I want either.

I think the only way for this to work is to display the picker first, wait until it's dismissed, and then to shown the keyboard. But I have no way currently to detect when the picker is dismissed.

Do you have any advice on how to detect when the picker is dismissed?

**Comment by lukmccall** [comment]
Hi @Ashoat, unfortunately, I think there is no official way to detect when or if the picker was dismissed. The functionality of "limited picker" is really restricted. Maybe there is a hacky solution to your problem, but I don't have any ideas.

**Comment by Ashoat** [comment]
Thanks for the response @lukmccall! I think my research corresponds to yours – Apple just doesn't support this right now for whatever reason. I guess the best thing I can do on my end is to give up on having a custom keyboard for the media gallery.
